filters/shedder/admission.go

```diff
@@ -0,0 +1,366 @@
+		}
+	}
+
+	maxRejectProbability := defaultMaxRejectProbability
+	if len(args) > 6 {
+		maxRejectProbability, err = getFloat64Arg(args[6])
+		if err != nil {
+			logrus.Warnf("maxRejectProbability failed: %v", err)
+			return nil, filters.ErrInvalidFilterParameters
+		}
+	}
+
+	aggression := defaultAggression
+	if len(args) > 7 {
+		aggression, err = getFloat64Arg(args[7])
+		if err != nil {
+			logrus.Warnf("aggression failed: %v", err)
+			return nil, filters.ErrInvalidFilterParameters
+		}
+		if aggression <= 0.0 {
+			logrus.Warn("aggression should be >0")
+			return nil, filters.ErrInvalidFilterParameters
+		}
+	}
+
+	ac := &admissionControl{
+		metrics:      metrics.Default,
+		metricSuffix: metricSuffix,
+		tracer:       spec.tracer,
+
+		mode:                 mode,
+		d:                    d,
+		windowSize:           windowSize,
+		minRequests:          minRequests,
+		successThreshold:     threshold,
+		maxRejectProbability: maxRejectProbability,
+		aggression:           aggression,
+
+		totals:         make([]int64, windowSize),
+		success:        make([]int64, windowSize),
+		counter:        atomic.NewInt64(0),
+		successCounter: atomic.NewInt64(0),
+	}
+	go ac.tickWindows()
+	return ac, nil
+}
+
+func (ac *admissionControl) tickWindows() {
+	t := time.NewTicker(ac.d)
+	i := 0
+	for range t.C {
+		val := ac.counter.Swap(0)
+		ok := ac.successCounter.Swap(0)
+
+		ac.mu.Lock()
+		ac.totals[i] = val
+		ac.success[i] = ok
+		ac.mu.Unlock()
+
+		i = (i + 1) % ac.windowSize
+	}
+}
+
+func (ac *admissionControl) count() (float64, float64) {
+	ac.mu.Lock()
+	defer ac.mu.Unlock()
+	return float64(sum(ac.totals)), float64(sum(ac.success))
+}
+
+func sum(a []int64) int64 {
+	var result int64
+	for _, v := range a {
+		result += v
+	}
+	return result
+}
+
+// calculates P_{reject} see https://opensource.zalando.com/skipper/reference/filters/#admissioncontrol
+func (ac *admissionControl) pReject() float64 {
+	var rejectP float64
+
+	total, success := ac.count()
+	factor := float64(ac.d) / float64(time.Second)
+	if total < float64(ac.minRequests)*factor {
+		return -1
+	}
+
+	s := success / ac.successThreshold
+	if ac.mode == log {
+		logrus.Infof("rejectP = (%0.2f - %0.2f) / (%0.2f + 1)  --- success: %0.2f and threshold: %0.2f", total, s, total, success, ac.successThreshold)
+	}
+	if total < s {
+		return -1
+	}
+	rejectP = (total - s) / (total + 1)
+	rejectP = math.Pow(rejectP, 1/ac.aggression)
+
+	rejectP = math.Min(rejectP, ac.maxRejectProbability)
+	return math.Max(rejectP, 0.0)
+}
+
+func (ac *admissionControl) Request(ctx filters.FilterContext) {
+	span := ac.startSpan(ctx.Request().Context())
+	defer span.Finish()
+
+	p := ac.pReject() // [0, ac.maxRejectProbability] and -1 to disable
+	/* #nosec */
+	r := rand.Float64() // [0,1)
+
+	if ac.mode == log {
+		logrus.Infof("p: %0.2f, r: %0.2f", p, r)
+	}
+
+	if p > r {
+		ac.metrics.IncCounter(counterPrefix + ac.metricSuffix)
+		ext.Error.Set(span, true)
```

Comment:
I think this should be an error only if we serve StatusServiceUnavailable.

Reply:
Maybe we also should tag the span with the mode if it is not `active`

Reply:
That's to enable visibility before we make it "active", but I will add 
```
span.SetTag("mode", ac.mode.String())
```
such that it's observable that we don't block it, otherwise we can not understand with "inactive" mode in opentracing the behavior.